libs/libc/net/lib_getifaddrs.c

```diff
@@ -0,0 +1,275 @@
+                     (FAR struct sockaddr *)&myaddrs->dstaddr;
+              memcpy(&myaddrs->dstaddr,
+                     &req.lifr_dstaddr, sizeof(req.lifr_dstaddr));
+            }
+          else if (ioctl(sockfd, SIOCGIFBRDADDR, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_broadaddr =
+                     (FAR struct sockaddr *)&myaddrs->broadaddr;
+              memcpy(&myaddrs->broadaddr,
+                     &req.lifr_broadaddr, sizeof(req.lifr_broadaddr));
+            }
+
+          if (ioctl(sockfd, SIOCGIFHWADDR, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_data = &myaddrs->hwaddr;
+              memcpy(&myaddrs->hwaddr,
+                     &req.lifr_hwaddr, sizeof(req.lifr_hwaddr));
+            }
+
+          myaddrs = NULL; /* Zero out to handle the dual stack */
+        }
+#endif
+
+#ifdef CONFIG_NET_IPv6
+      if (ioctl(sockfd, SIOCGLIFADDR, (unsigned long)&req) >= 0)
+        {
+          if (myaddrs == NULL)
+            {
+              myaddrs->addrs.ifa_next = lib_zalloc(sizeof(*myaddrs));
+              myaddrs = (FAR struct myifaddrs *)myaddrs->addrs.ifa_next;
+
+              if (myaddrs == NULL)
+                {
+                  goto err;
+                }
+
+              myaddrs->addrs.ifa_name = myaddrs->name;
+              strncpy(myaddrs->name, req.lifr_name, IF_NAMESIZE);
+
+              myaddrs->addrs.ifa_flags = flags;
+            }
+
+          myaddrs->addrs.ifa_addr = (FAR struct sockaddr *)&myaddrs->addr;
+          memcpy(&myaddrs->addr, &req.lifr_addr, sizeof(req.lifr_addr));
+
+          if (ioctl(sockfd, SIOCGLIFNETMASK, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_netmask =
+                     (FAR struct sockaddr *)&myaddrs->netmask;
+              memcpy(&myaddrs->netmask,
+                     &req.lifr_netmask, sizeof(req.lifr_netmask));
+            }
+
+          if (ioctl(sockfd, SIOCGLIFDSTADDR, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_dstaddr =
+                     (FAR struct sockaddr *)&myaddrs->dstaddr;
+              memcpy(&myaddrs->dstaddr,
+                     &req.lifr_dstaddr, sizeof(req.lifr_dstaddr));
+            }
+          else if (ioctl(sockfd, SIOCGLIFBRDADDR, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_broadaddr =
+                     (FAR struct sockaddr *)&myaddrs->broadaddr;
+              memcpy(&myaddrs->broadaddr,
+                     &req.lifr_broadaddr, sizeof(req.lifr_broadaddr));
+            }
+
+          if (ioctl(sockfd, SIOCGIFHWADDR, (unsigned long)&req) >= 0)
+            {
+              myaddrs->addrs.ifa_data = &myaddrs->hwaddr;
+              memcpy(&myaddrs->hwaddr,
+                     &req.lifr_hwaddr, sizeof(req.lifr_hwaddr));
+            }
+        }
+#endif
+    }
+
+  close(sockfd);
+  return OK;
+
+err:
+  freeifaddrs(*addrs);
```

Comment:
```suggestion
  if (*addrs)
  {
    freeifaddrs(*addrs);
  }
```

Reply:
This is because
```
      if (ioctl(sockfd, SIOCGIFFLAGS, (unsigned long)&req) < 0)
        {
          goto err;
        }
```
may happen while `*addrs` not allocated.

Reply:
Don't need since freeifaddrs check null before doing the real action:
```
void freeifaddrs(FAR struct ifaddrs *addrs)
{
  while (addrs != NULL)
    {
      FAR struct ifaddrs *tmp = addrs;
      addrs = addrs->ifa_next;
      lib_free(tmp);
    }
}
```

Reply:
Anyway, I add the check and reset to NULL in case of fail.

Reply:
Yes. I missed that `addrs` is checked inside the `freeifaddrs`. The fix is actually to init `*addrs` to `NULL` before the loop.